src/lib.rs

```diff
@@ -96,20 +97,146 @@ pub struct SelectionOutput {
+        &options,
+        rng,
+    );
+    match bnb_selected_coin {
+        Some(selected_coin) => {
+            let accumulated_value: u64 = selected_coin
+                .iter()
+                .fold(0, |acc, &i| acc + inputs[i].value);
+            let accumulated_weight: u32 = selected_coin
+                .iter()
+                .fold(0, |acc, &i| acc + inputs[i].weight);
+            let estimated_fee = 0;
+            let waste = calculate_waste(
+                inputs,
+                &selected_inputs,
+                &options,
+                accumulated_value,
+                accumulated_weight,
+                estimated_fee,
+            );
+            let selection_output = SelectionOutput {
+                selected_inputs: selected_coin,
+                waste: WasteMetric(waste),
+            };
+            Ok(selection_output)
+        }
+        None => select_coin_srd(inputs, options, &mut rand::thread_rng()),
+    }
 }
 
 /// Return empty vec if no solutions are found
+// changing the selected_inputs : &[usize] -> &mut Vec<usize>
 fn bnb(
     inputs_in_desc_value: &[(usize, OutputGroup)],
-    selected_inputs: &[usize],
-    effective_value: u64,
+    selected_inputs: &mut Vec<usize>,
+    acc_eff_value: u64,
     depth: usize,
     bnp_tries: u32,
     options: &CoinSelectionOpt,
-) -> Vec<usize> {
-    unimplemented!()
+    rng: &mut ThreadRng,
+) -> Option<Vec<usize>> {
+    let target_for_match = options.target_value
+        + calculate_fee(options.base_weight, options.target_feerate)
+        + options.cost_per_output;
+    let match_range = options.cost_per_input + options.cost_per_output;
+    if acc_eff_value > target_for_match + match_range {
+        return None;
+    } else if acc_eff_value >= target_for_match {
+        return Some(selected_inputs.to_vec());
+    } else if bnp_tries <= 0 || depth >= inputs_in_desc_value.len() {
+        return None;
+    } else {
+        if rng.gen_bool(0.5) {
+            // exploring the inclusion branch
+            // first include then omit
+            let new_effective_values = acc_eff_value
+                + effective_value(&inputs_in_desc_value[depth].1, options.target_feerate);
+            selected_inputs.push(inputs_in_desc_value[depth].0);
+            let with_this = bnb(
+                inputs_in_desc_value,
+                selected_inputs,
+                new_effective_values,
+                depth + 1,
+                bnp_tries - 1,
+                options,
+                rng,
+            );
+            match with_this {
+                Some(_) => return with_this,
+                None => {
+                    selected_inputs.pop(); //poping out the selected utxo if it does not fit
+                    let without_this = bnb(
+                        inputs_in_desc_value,
+                        selected_inputs,
+                        acc_eff_value,
+                        depth + 1,
+                        bnp_tries - 1,
```

Comment:
bnp_tries - 2

Reply:
because every recursive call should decrement this by one.
This variable is immutable